include/list.h

```diff
@@ -0,0 +1,431 @@
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdint.h>
+#include <stdbool.h>
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/**
+ * @struct list_desc
+ * @brief Structure storing the list and function wrapper for usual list types
+ *
+ * With this structure the funtionalities of usual list types
+ * ( \ref adapter_type ) can be accesed with the functions referenced in this
+ * structure.
+ * For example:
+ * @code{.c}
+ * my_queue->push(my_queue, my_data);
+ * @endcode
+ */
+struct list_desc;
+
+/**
+ * @struct list_iterator
+ * @brief Structure used to iterate through the list using Iterator functions.
+ */
+struct iterator;
+
+/**
+ * @brief Prototype of the compare function.
+ *
+ * The function used to compare the elements of the liste when doing
+ * operations on an ordered list.
+ * @param data1 - First element to be compared
+ * @param data2 - Second element to be compared
+ * @return
+ *  - -1 - If data1 < data2
+ *  - 0 - If data1 == data2
+ *  - 1 - If data1 > data2
+ */
+typedef int32_t (*f_cmp)(void *data1, void *data2);
+
+/**
+ * @name Generic functions
+ * Each function interacting with the list have one of the following formats.\n
+ * Aditionaly they may have one more parametere for specific functionalities.\n
+ * In the Iterator functions, the list reference is replaced by the iterator's
+ * one.
+ * @{
+ */
+
+/**
+ * @brief Add an element in the list.
+ *
+ * The element of the list is created and the data field is stored in it.
+ * @param list_desc - Reference to the list. Created by \ref list_init
+ * @param data - Data to store in a list element
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+typedef int32_t (*f_add)(struct list_desc *list_desc, void *data);
+
+/**
+ * @brief Edit an element in the list. The content is replaced by new_data.
+ * @param list_desc - Reference to the list. Created by \ref list_init .
+ * @param new_data - New data to replace the old one
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+typedef int32_t (*f_edit)(struct list_desc *list_desc, void *new_data);
+
+/**
+ * @brief Read an element from the list.
+ * @param list_desc - Reference to the list. Created by \ref list_init
+ * @param result - If not null, result is filled with:
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ * @return \n
+ *  - Content of the list element
+ *  - NULL : If some error occure
+ * @note If the content of an element can be 0 then the result must be checked
+ * to see if the functions has succeded
+ */
+typedef void* (*f_read)(struct list_desc *list_desc, int32_t *result);
```

Comment:
Usually we use the return of the function to check the status and an output parameter to get something out of the function. Maybe we should keep the consistency on this matter.

Reply:
First I tough to write them as you are saying. Then I tough it will be more practical this way.
you can use it as function parameter for example:
`printf("%d", queue->pop(queue, NULL));`
Assuming the function will fail only I the list is not handled properly, which can be check during development using the result param.
If after this explanation you think I should change it, I will change it.

Reply:
If the function can fail at any point, the return value should be an error code and any output parameter should be passed back by pointers. Rephrased, the only situation where the return value could be a meaningful output parameter is when there is no way the function can fail.

I missed this during my review, but I think what @adrimbarean says is the correct approach here since this function has failure points.